### ISSUE TITLE
Redesign function signature data structures

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/syn#595]
As of Syn 0.15 we currently have two structs responsible for function signatures:

```rust
struct MethodSig {
    constness: Option<Token![const]>,
    unsafety: Option<Token![unsafe]>,
    asyncness: Option<Token![async]>,
    abi: Option<Abi>,
    ident: Ident,
    decl: FnDecl,
}
```

```rust
struct FnDecl {
    fn_token: Token![fn],
    generics: Generics,
    paren_token: token::Paren,
    inputs: Punctuated<FnArg, Token![,]>,
    variadic: Option<Token![...]>,
    output: ReturnType,
}
```

with no clear distinction between what they mean, other than one containing the other.

They are used from the rest of the syntax tree as follows:

```rust
struct TraitItemMethod {
    attrs: Vec<Attribute>,
    sig: MethodSig,
    default: Option<Block>,
    semi_token: Option<Token![;]>,
}
```

```rust
struct ImplItemMethod {
    attrs: Vec<Attribute>,
    vis: Visibility,
    defaultness: Option<Token![default]>,
    sig: MethodSig,
    block: Block,
}
```

```rust
struct ItemFn {
    attrs: Vec<Attribute>,
    vis: Visibility,
    constness: Option<Token![const]>,
    unsafety: Option<Token![unsafe]>,
    asyncness: Option<Token![async]>,
    abi: Option<Abi>,
    ident: Ident,
    decl: Box<FnDecl>,
    block: Box<Block>,
}
```

```rust
struct ForeignItemFn {
    attrs: Vec<Attribute>,
    vis: Visibility,
    ident: Ident,
    decl: Box<FnDecl>,
    semi_token: Token![;],
}
```

This setup faithfully represents the current syntax restrictions of Rust.

- Trait methods cannot have a visibility.
- Only impl methods can have `default`.
- Foreign functions cannot have `const` / `unsafe` / `async` / abi. This appears to account for the split between MethodSig and FnDecl.

I think it would be fine for Syn to bend the last restriction and have a single function signature data type that is shared among all four of TraitItemMethod, ImplItemMethod, ItemFn, ForeignItemFn.

```rust
struct Signature {
    constness: Option<Token![const]>,
    unsafety: Option<Token![unsafe]>,
    asyncness: Option<Token![async]>,
    abi: Option<Abi>,
    fn_token: Token![fn],
    ident: Ident,
    generics: Generics,
    paren_token: token::Paren,
    inputs: Punctuated<FnArg, Token![,]>,
    variadic: Option<Token![...]>,
    output: ReturnType,
}
```

**Comment by Boiethios** [comment]
Just a little message to say that I would be happy to see that. In one of my projects, I must treat a method and a function more or less the same, but to do so I must do some contortions in my code.

Thanks for this great crate, BTW.

**Comment by dtolnay** [comment]
Implemented in c7369974dfaf3e79a7c809e35127276ff6326049.
